File: Two Pointers/Make Lexicographically Smallest Array by Swapping Elements/Make_Lexicographically_Smallest_Array_by_Swapping_Elements.cpp
// Problem Name: Make Lexicographically Smallest Array by Swapping Elements
// Problem Link: https://leetcode.com/problems/make-lexicographically-smallest-array-by-swapping-elements/description/ 

class Solution {
public:
    vector<int> lexicographicallySmallestArray(vector<int>& nums, int limit) {
        int n = nums.size();
        vector<pair<int, int>> vp;
        for (int i = 0; i < n; i++) {
            vp.push_back({nums[i], i});
        }
        sort(vp.begin(), vp.end());
        int left = 0, right = 1;
        while (right < n) {
            vector<int> pos = {vp[left].second};
            while (right < n &&
                   abs(vp[right].first - vp[right - 1].first) <= limit) {
                pos.push_back(vp[right].second);
                right++;
            }

            sort(pos.begin(), pos.end());
            for (int i = 0; i < right - left; i++) {
                nums[pos[i]] = vp[left + i].first;
            }
            left = right;
            right++;
        }
        return nums;
    }
};

File: Two Pointers/Make Lexicographically Smallest Array by Swapping Elements/Make_Lexicographically_Smallest_Array_by_Swapping_Elements.java
// Problem Name: Make Lexicographically Smallest Array by Swapping Elements
// Problem Link: https://leetcode.com/problems/make-lexicographically-smallest-array-by-swapping-elements/description/ 

import java.util.*;

class Solution {
    public int[] lexicographicallySmallestArray(int[] nums, int limit) {
        int n = nums.length;
        List<int[]> copy = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            copy.add(new int[] { nums[i], i });
        }
        copy.sort(Comparator.comparingInt(a -> a[0]));
        int left = 0, right = 1;
        while (right < n) {
            List<Integer> pos = new ArrayList<>();
            pos.add(copy.get(left)[1]);
            while (right < n && Math.abs(copy.get(right)[0] - copy.get(right - 1)[0]) <= limit) {
                pos.add(copy.get(right)[1]);
                right++;
            }

            Collections.sort(pos);
            for (int i = 0; i < right - left; i++) {
                nums[pos.get(i)] = copy.get(left + i)[0];
            }
            left = right;
            right++;
        }
        return nums;
    }
}